File: myproj26/main12-high-order-function.py
def base_10(fn): # base_10함수가 호출이 되면 
    def wrap(x,y): # 매번 새로운 wrap라는 함수를 만들어서 리턴
        return fn(x,y) +10
    return wrap
# 이렇게 함수를 인자로 받고, 
# 내부적으로 함수를 만들어서 리턴할 수 있음 
# 이렇게 하는 것이 파이썬의 장식자
# Decorators 

# 1 번
@base_10 
@base_10 # 이렇게 두번 할 수도 있음 -> 그러면 20이 깔림 
def mysum(x,y):
    return x+y

# 2번 (1번, 2번은 같은 것! )
mysum = base_10(base_10(mysum))
print(mysum(1,2))

# 이렇게 동일 함수로 두번 감아쓸 수 있음

File: myproj26/main12-high-order-function.js
// 파이썬의 장식자 문법과 비슷 
// 
// function base_10(fn) {
//     function wrap(x, y) {
//         return fn(x, y) + 10;
//     }
//     return wrap;
// }


// function mysum(x, y) {
//     return x + y;
// }

// mysum = base_10(base_10(mysum));

// console.log(mysum(1, 2));




// arrow function 버전으로 바꾸기 



// async function wrap1(x, y) {
//     //에러는?
//     try { // 한줄 읽어내고, 이줄에서 에러가 발생하면 에러 코드로 넘어가고 
//         function fn(x, y) {
//             return fn(x, y) + 10
//         };

//     } // 비동기지만, 동기적인 코드를 작성하는 것을 도와줌 
//     catch (error) {
//         console.error(error);
//     }

// }
// console.log(wrap1(1, 2));


/// arrow function
const base_10 = (fn) => (x, y) => fn(x, y) + 10;

// 이 코드를 풀면?
// function base_10(fn) {
//     function wrap(x, y) {
//         return fn(x, y) + 10;
//     }
//     return wrap;
// } -> 이거랑 같음


// wrap 만들자 마자 리턴 -> 

